pkg/proxy/proxy_community_test.go

```diff
@@ -0,0 +1,306 @@
+		}))
+		defer testServer.Close()
+
+		type testCase struct {
+			ProxyRequestMethod              string
+			ProxyRequestHeaders             http.Header
+			ExpectedAPIServerRequestHeaders http.Header
+			ExpectedProxyResponseStatus     int
+			RequestPath                     string
+			ExpectedResponse                string
+		}
+
+		tests := map[string]testCase{
+			// Given smith2 owns a workspace named communityspace
+			// And   communityspace is publicly visible (shared with PublicViewer)
+			// When  smith2 requests the list of pods in workspace communityspace
+			// Then  the request is forwarded from the proxy
+			// And   the request impersonates smith2
+			// And   the request's X-SSO-User Header is set to smith2's ID
+			// And   the request is successful
+			"plain http actual request as owner": {
+				ProxyRequestMethod:  "GET",
+				ProxyRequestHeaders: map[string][]string{"Authorization": {"Bearer " + s.token(owner)}},
+				ExpectedAPIServerRequestHeaders: map[string][]string{
+					"Authorization":    {"Bearer clusterSAToken"},
+					"Impersonate-User": {"smith2"},
+					"X-SSO-User":       {"username-" + owner.String()},
+				},
+				ExpectedProxyResponseStatus: http.StatusOK,
+				RequestPath:                 fmt.Sprintf("http://localhost:%s/workspaces/communityspace/api/communityspace/pods", port),
+				ExpectedResponse:            httpTestServerResponse,
+			},
+			// Given A not ready user exists
+			// When  the not ready user requests the list of pods in workspace communityspace
+			// Then  the request is forwarded from the proxy
+			// And   the request impersonates the not ready user
+			// And   the request's X-SSO-User Header is set to not ready user's ID
+			// And   the request is successful
+			"plain http actual request as notReadyUser": {
+				ProxyRequestMethod:  "GET",
+				ProxyRequestHeaders: map[string][]string{"Authorization": {"Bearer " + s.token(notReadyUser)}},
+				ExpectedAPIServerRequestHeaders: map[string][]string{
+					"Authorization":    {"Bearer clusterSAToken"},
+					"Impersonate-User": {toolchainv1alpha1.KubesawAuthenticatedUsername},
+					"X-SSO-User":       {"username-" + notReadyUser.String()},
+				},
+				ExpectedProxyResponseStatus: http.StatusOK,
+				RequestPath:                 fmt.Sprintf("http://localhost:%s/workspaces/communityspace/api/communityspace/pods", port),
+				ExpectedResponse:            httpTestServerResponse,
+			},
+			// Given smith2 owns a workspace named communityspace
+			// And   communityspace is publicly visible (shared with PublicViewer)
+			// And   a user named communityuser exists
+			// And   smith2's communityspace is not directly shared with communityuser
+			// When  communityuser requests the list of pods in workspace communityspace
+			// Then  the request is forwarded from the proxy
+			// And   the request impersonates the PublicViewer
+			// And   the request's X-SSO-User Header is set to communityuser's ID
+			// And   the request is successful
```

Comment:
The ` communityuser` name is a bit confusing. It's just a regular user, right? There is nothing special in this user.
I would maybe use the following setup:
- User `smith` with publicly available workspace `smith-communityspace`
- User `alice` with private workspace `alice-private`
- User `bob` with private workspace `bob-private`
- User `john-not-ready` (not-ready)

Then check that:
- `smith`  and `john-notready` can access `smith-communityspace` only.
- `alice` can access `alice-private` and `smith-communityspace`
- `bob` can access `bob-private` and `smith-communityspace`
- some non-existing user can access `smith-communityspace` only

That would cover all use cases.
And I would also add a test for banned user after the banned user middleware PR is merged. We need that PR get merged first anyway.

Reply:
Thanks for the suggestion, I cleaned it up a bit and used the following users:
- User `smith` with publicly available workspace `smith-community`
- User `alice` with private workspace `alice-private`
- User `bob` as the non-signed-up user
- User `john` as the not-ready user

Now we check:
- `smith` can access `smith-community` as `smith`
- `john` can access `smith-community` as `kubesaw-authenticated`
- `bob` can access `smith-community` as `kubesaw-authenticated`
- `alice` can access `smith-community` as `kubesaw-authenticated`
- `smith` can not access `alice-private`

I'll add the test case for banned user

Reply:
added in ea515a6: `eve` is the banned user